torch/_inductor/runtime/triton_bundler.py

```diff
@@ -0,0 +1,127 @@
+import dataclasses
+import logging
+import os
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import torch
+from torch._dynamo.utils import counters
+
+from .runtime_utils import cache_dir
+
+
+log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundleEntry:
+    kernel_hash: str
+    device: int
+    directory: str
+
+
+@dataclasses.dataclass
+class TritonKernelArtifact:
+    filename: str
+    payload: bytes = dataclasses.field(repr=False)  # Do not display binary
+
+
+@dataclasses.dataclass
+class TritonKernelArtifacts:
+    kernel_hash: str
+    device: int
+    artifacts: List[TritonKernelArtifact]
+
+
+class TritonBundler:
+    # It is safe for multiple threads to insert entries as insert operation
+    # of dict is atomic
+    _entries: Optional[Dict[TritonBundleEntry, None]] = None
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
+
+        if not config.is_fbcode():
+            return False
+
+        return torch._utils_internal.justknobs_check(
+            "pytorch/remote_cache:bundle_triton_into_fx_graph_cache"
+        )
+
+    @classmethod
+    def begin_compile(cls) -> None:
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = {}
+
+    @classmethod
+    def put(cls, kernel_hash: str, device: int) -> None:
+        if (entries := cls._entries) is not None:
+            # CachingAutotuner.__init__ unconditionally sets TRITON_CACHE_DIR
+            triton_cache_dir = os.getenv("TRITON_CACHE_DIR")
+            assert triton_cache_dir is not None
+            entries[TritonBundleEntry(kernel_hash, device, triton_cache_dir)] = None
+
+    @classmethod
+    def collect(cls) -> List[TritonKernelArtifacts]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
+            return []
+
+        if (entries := cls._entries) is not None:
```

Comment:
Should we assert that cls._entries it not None here? The begin compile method seems to indicate that we're trying to ensure one compile at a time, but we don't check that any entries are written here.

Or are we instead going to sometimes have to _entries even for valid compiles so we can't assert correctness here?

Reply:
I had this it this way because what if someone enables just knobs in the middle of execution, so entries is None but collect operation executes because it was turned on midway

Reply:
Hmmm - I guess I'm curious about if it makes sense to have begin_compile around at all? just set _entires in put, read + clear them in collect?  But very much your call.

Seperately if JK is turned on between begin_compile + put, we're going to not write to the filesystem during put, not sure if that's an issue (if inductor is going to read the files later).